casper/src/main/scala/coop/rchain/casper/engine/GenesisValidator.scala

```diff
@@ -30,13 +40,26 @@ class GenesisValidator[F[_]: Sync: Metrics: Span: Concurrent: CommUtil: Transpor
   private val F    = Applicative[F]
   private val noop = F.unit
 
+  implicit val state = Cell.unsafe[F, Map[BlockHash, Boolean]](Map.empty)
```

Comment:
Implicits are powerful abstraction. If it's not necessary, it can make code harder to reason about. `seen` and `ack` can be local functions accessing `state`.
E.g. as mentioned in the conclusion for implicit functions _please keep the Principle of Least Power in mind_.
https://www.scala-lang.org/blog/2016/12/07/implicit-function-types.html#conclusion

By the way, implicit functions look very nice for something like program configuration. I'm not sure if they will be available only in Dotty?

Reply:
I'm not sure I follow the gist about configuration. Do you mean CLI options, config files etc?

Reply:
@tgrospic https://github.com/rchain/rchain/pull/2910/commits/68993b6c159adb4dce40eeeed137d8f2d519a144 how about this?

Reply:
Excellent!

I was referring to the use of implicits for configuration.
> In a functional setting, the inputs to a computation are most naturally expressed as parameters. One could simply augment functions to take additional parameters that represent configurations, capabilities, dictionaries, or whatever contextual data the functions need.

I remember you question, where should be the default values for configuration. It depends from where and in which moment someone is asking. Implementation of _Config_ implicit can load it from DB or expose mutation to the API.